<a href="https://colab.research.google.com/github/Sanjay-sundaresan/ML-Project/blob/main/ML_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Relationship between imbalance and overlapping of datasets - Imbalance and overlapping are considered together and treated as a single issue. 

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import MinMaxScaler
from sklearn.model_selection import KFold 
from sklearn.metrics import classification_report
from sklearn import metrics
from sklearn import svm 
from sklearn.model_selection import cross_val_score
from numpy import mean

In [ ]:
from google.colab import drive
drive.mount("/content/gdrive")

Mounted at /content/gdrive


Reading Synthetically generated Datasets

In [ ]:
unbalanced_3F_overlap = pd.read_csv("/content/gdrive/My Drive/ML_DATA/unbalanced-3-F-overlaple.csv")

In [ ]:
unbalanced_allF_overlap = pd.read_csv("/content/gdrive/My Drive/ML_DATA/unbalanced-all-F-overlap.csv")

In [ ]:
unbalanced_noF_overlap = pd.read_csv("/content/gdrive/My Drive/ML_DATA/unbalanced-no-F-overlap.csv")

In [ ]:
unbalanced_1F_overlap = pd.read_csv("/content/gdrive/My Drive/ML_DATA/unbalanced-one-F-overlap.csv")

In [ ]:
Balanced_2F_overlap = pd.read_csv("/content/gdrive/My Drive/ML_DATA/Balanced-2-F-overlap.csv")

In [ ]:
Balanced_3F_overlap = pd.read_csv("/content/gdrive/My Drive/ML_DATA/Balanced-3-F-overlap.csv")

In [ ]:
Balanced_AllF_overlap = pd.read_csv("/content/gdrive/My Drive/ML_DATA/Balanced-All-F-overlap.csv")

In [ ]:
Balanced_noF_overlap = pd.read_csv("/content/gdrive/My Drive/ML_DATA/Balanced-no-F-overlap.csv")

In [ ]:
Balanced_oneF_overlap = pd.read_csv("/content/gdrive/My Drive/ML_DATA/Balanced-one-F-overlap.csv")

1. Consistency analysis of balanced dataset with all 4 features overlapping

In [ ]:
Balanced_AllF_overlap.describe()

X1          X2          X3          X4           Y
count  300.000000  300.000000  300.000000  300.000000  300.000000
mean     0.984215    0.997026    0.975882    1.009829    0.500000
std      0.269708    0.218569    0.238782    0.214174    0.500835
min      0.112605    0.361477    0.275235    0.297676    0.000000
25%      0.817970    0.872916    0.831546    0.885946    0.000000
50%      0.992419    0.978608    0.982293    1.007222    0.500000
75%      1.155689    1.115385    1.143743    1.136795    1.000000
max      1.817953    1.821522    1.594692    1.615540    1.000000

In [ ]:
Balanced_AllF_overlap.skew()

X1   -0.080456
X2    0.265152
X3   -0.156417
X4   -0.159887
Y     0.000000
dtype: float64

2. Consistency analysis of **unbalanced** dataset with all 4 features overlapping

In [ ]:
unbalanced_allF_overlap.describe()

X1          X2          X3          X4           Y
count  300.000000  300.000000  300.000000  300.000000  300.000000
mean     0.818889    1.221707    0.885161    0.813715    0.800000
std      0.239075    0.228633    0.258988    0.260760    0.400668
min      0.080868    0.548849    0.224936    0.020823    0.000000
25%      0.672422    1.070023    0.706794    0.625986    1.000000
50%      0.830488    1.229283    0.855104    0.810631    1.000000
75%      0.978032    1.363268    1.036426    0.991449    1.000000
max      1.538569    1.912698    1.884021    1.728664    1.000000

In [ ]:
unbalanced_allF_overlap.skew()

X1   -0.137918
X2    0.008775
X3    0.496669
X4    0.074451
Y    -1.507548
dtype: float64

The skewness of the features in both cases indicates that the class boundaries are not very distinct, thus making classification more challenging. 

3. Principal component analysis of balanced dataset with all features overlapping. 

In [ ]:
features = ['X1','X2','X3','X4']
x = Balanced_AllF_overlap.loc[:, features].values
x = StandardScaler().fit_transform(x)
pd.DataFrame(data = x, columns = features).head()

X1        X2        X3        X4
0  0.800117  1.196404  0.029285 -0.044495
1  0.814772  0.318165 -1.850527 -0.170702
2  0.416745  0.859367  1.333230  1.736665
3 -0.096859 -0.428794 -0.238741 -1.314712
4  0.839079  0.741537  0.775930  1.337607

In [ ]:
pca = PCA(n_components= 4)
principalComponents = pca.fit_transform(x)
pca_df = pd.DataFrame(data = principalComponents, columns = ['Component 1', 'Component 2', 'Component 3', 'Component 4'])
pca_df.head()

Component 1  Component 2  Component 3  Component 4
0     1.072513    -0.800458     0.495719     0.194035
1    -0.445672    -1.199168     1.354960    -0.863787
2     2.074759     1.177914     0.103649    -0.053349
3    -0.931819    -0.853194    -0.555431     0.271996
4     1.774130     0.523916     0.237242    -0.408513

In [ ]:
B = pd.DataFrame(["Component 1", "Component 2","Component 3","Component 4"], columns = ["Component"])
Bdf = pd.DataFrame(pca.explained_variance_ratio_ * 100, columns= ["Perc_of_Variance"])
Bdf = pd.concat([B, Bdf], axis = 1)
Bd = Bdf.Perc_of_Variance.values
Bd1= [0,0,0,0]
for i in range(0,4):
  if i == 0 :
    Bd1[i] = Bd[i]
  else:
    Bd1[i] = Bd[i] + Bd1[i-1] 
Bdf['Cumulative_Perc'] = Bd1
Bdf.head()

Component  Perc_of_Variance  Cumulative_Perc
0  Component 1         60.658915        60.658915
1  Component 2         25.328284        85.987198
2  Component 3         10.959721        96.946920
3  Component 4          3.053080       100.000000

4. Principal component analysis of **unbalanced** dataset with all features overlapping. 

In [ ]:
features = ['X1','X2','X3','X4']
x = unbalanced_allF_overlap.loc[:, features].values
x = StandardScaler().fit_transform(x)
pd.DataFrame(data = x, columns = features).head()

X1        X2        X3        X4
0 -0.914255  1.503730  0.367788 -0.024323
1 -1.755542  1.758760  1.502711 -1.733837
2 -0.584866  0.035396  2.460938  0.211248
3  1.208885 -0.158833  2.009902  0.690207
4  0.156440  0.772606  1.663667  0.602989

In [ ]:
pca = PCA(n_components= 4)
principalComponents = pca.fit_transform(x)
pca_df = pd.DataFrame(data = principalComponents, columns = ['Component 1', 'Component 2', 'Component 3', 'Component 4'])
pca_df.head()

Component 1  Component 2  Component 3  Component 4
0    -0.679441     1.051787    -0.308369    -1.252971
1    -1.330368     3.057343    -0.055368    -0.564986
2     0.631596     1.756607    -1.682542     0.358651
3     0.258102     0.059625    -2.239620     0.957418
4    -0.077924     0.727596    -1.782730    -0.197941

In [ ]:
B = pd.DataFrame(["Component 1", "Component 2","Component 3","Component 4"], columns = ["Component"])
Bdf = pd.DataFrame(pca.explained_variance_ratio_ * 100, columns= ["Perc_of_Variance"])
Bdf = pd.concat([B, Bdf], axis = 1)
Bd = Bdf.Perc_of_Variance.values
Bd1= [0,0,0,0]
for i in range(0,4):
  if i == 0 :
    Bd1[i] = Bd[i]
  else:
    Bd1[i] = Bd[i] + Bd1[i-1] 
Bdf['Cumulative_Perc'] = Bd1
Bdf.head()

Component  Perc_of_Variance  Cumulative_Perc
0  Component 1         36.107030        36.107030
1  Component 2         28.894966        65.001995
2  Component 3         24.916998        89.918993
3  Component 4         10.081007       100.000000

For balanced data, the first component explains ~60.66% of the
total variance, the second component explains ~25.33%, and the fourth one represents ~3.05% of the variance. The PCA for unbalanced dataset, however, indicated that the first PCA component explains only ~36.1% of the variance, the second one ~28.9%, and the fourth one accounted for 10.1% of
the variance. This provides an idea about the difference in the distribution of the balanced and unbalanced sets with full overlapping. 

**Using 10-fold Cross-Validation to evaluate the models and ultimately compare the performance of different datasets.**

In [ ]:
Kfold = KFold(10, True)
overall_acc = []
overall_prec = []
overall_rec = []
sc = []

5. **Running SVM classifier using 10-fold cross validation for balanced and unbalanced datasets with varying degrees of overlapping**

5.1: **Balanced dataset with 1 feature overlap**

In [ ]:
accuracies = []
precisions = []
recalls = []

features = ['X1','X2','X3','X4']
x = Balanced_oneF_overlap.loc[:, features].values
x = pd.DataFrame(data = x, columns = features)
y = Balanced_oneF_overlap.Y
scaler = MinMaxScaler(feature_range=(0, 1))
x = scaler.fit_transform(x)

In [ ]:
clf = svm.SVC()
for train_index, test_index in Kfold.split(x):
  x_train, x_test, y_train, y_test = x[train_index], x[test_index], y[train_index], y[test_index]
  clf.fit(x_train, y_train)
  pred = clf.predict(x_test)
  accuracies.append(metrics.accuracy_score(y_test, pred))
  precisions.append(metrics.precision_score(y_test, pred))
  recalls.append(metrics.recall_score(y_test, pred))
print("Accuracy:",mean(accuracies))
print("Precision:",mean(precisions))
print("Recall:",mean(recalls))
overall_acc.append(mean(accuracies) * 100)
overall_prec.append(mean(precisions) * 100)
overall_rec.append(mean(recalls) * 100)
sc.append("Balanced-1-F-overlap")


Accuracy: 1.0
Precision: 1.0
Recall: 1.0


5.2: **Unbalanced dataset with 1 feature overlap**

In [ ]:
accuracies = []
precisions = []
recalls = []

features = ['X1','X2','X3','X4']
x = unbalanced_1F_overlap.loc[:, features].values
x = pd.DataFrame(data = x, columns = features)
y = unbalanced_1F_overlap.Y
scaler = MinMaxScaler(feature_range=(0, 1))
x = scaler.fit_transform(x)

In [ ]:
clf = svm.SVC()
for train_index, test_index in Kfold.split(x):
  x_train, x_test, y_train, y_test = x[train_index], x[test_index], y[train_index], y[test_index]
  clf.fit(x_train, y_train)
  pred = clf.predict(x_test)
  accuracies.append(metrics.accuracy_score(y_test, pred))
  precisions.append(metrics.precision_score(y_test, pred))
  recalls.append(metrics.recall_score(y_test, pred))
print("Accuracy:",mean(accuracies))
print("Precision:",mean(precisions))
print("Recall:",mean(recalls))
overall_acc.append(mean(accuracies) * 100)
overall_prec.append(mean(precisions) * 100)
overall_rec.append(mean(recalls) * 100)
sc.append("Unbalanced-1-F-overlap")

Accuracy: 1.0
Precision: 1.0
Recall: 1.0


5.3: **Balanced dataset with 2 feature overlap**

In [ ]:
accuracies = []
precisions = []
recalls = []

features = [' X1','X2','X3','X4']
x = Balanced_2F_overlap.loc[:, features].values
x = pd.DataFrame(data = x, columns = features)
y = Balanced_2F_overlap.Y
scaler = MinMaxScaler(feature_range=(0, 1))
x = scaler.fit_transform(x)

In [ ]:
clf = svm.SVC()
for train_index, test_index in Kfold.split(x):
  x_train, x_test, y_train, y_test = x[train_index], x[test_index], y[train_index], y[test_index]
  clf.fit(x_train, y_train)
  pred = clf.predict(x_test)
  accuracies.append(metrics.accuracy_score(y_test, pred))
  precisions.append(metrics.precision_score(y_test, pred))
  recalls.append(metrics.recall_score(y_test, pred))
print("Accuracy:",mean(accuracies))
print("Precision:",mean(precisions))
print("Recall:",mean(recalls))
overall_acc.append(mean(accuracies) * 100)
overall_prec.append(mean(precisions) * 100)
overall_rec.append(mean(recalls) * 100)
sc.append("Balanced-2-F-overlap")

Accuracy: 1.0
Precision: 1.0
Recall: 1.0


5.4: **Balanced dataset with 3 feature overlap**

In [ ]:
accuracies = []
precisions = []
recalls = []

features = ['X1','X2','X3','X4']
x = Balanced_3F_overlap.loc[:, features].values
x = pd.DataFrame(data = x, columns = features)
y = Balanced_3F_overlap.Y
scaler = MinMaxScaler(feature_range=(0, 1))
x = scaler.fit_transform(x)

In [ ]:
clf = svm.SVC()
for train_index, test_index in Kfold.split(x):
  x_train, x_test, y_train, y_test = x[train_index], x[test_index], y[train_index], y[test_index]
  clf.fit(x_train, y_train)
  pred = clf.predict(x_test)
  accuracies.append(metrics.accuracy_score(y_test, pred))
  precisions.append(metrics.precision_score(y_test, pred))
  recalls.append(metrics.recall_score(y_test, pred))
print("Accuracy:",mean(accuracies))
print("Precision:",mean(precisions))
print("Recall:",mean(recalls))
overall_acc.append(mean(accuracies) * 100)
overall_prec.append(mean(precisions) * 100)
overall_rec.append(mean(recalls) * 100)
sc.append("Balanced-3-F-overlap")

Accuracy: 1.0
Precision: 1.0
Recall: 1.0


5.5: **Unbalanced dataset with 3 feature overlap**

In [ ]:
accuracies = []
precisions = []
recalls = []

features = ['X1','X2','X3','X4']
x = unbalanced_3F_overlap.loc[:, features].values
x = pd.DataFrame(data = x, columns = features)
y = unbalanced_3F_overlap.Y
scaler = MinMaxScaler(feature_range=(0, 1))
x = scaler.fit_transform(x)

In [ ]:
clf = svm.SVC()
for train_index, test_index in Kfold.split(x):
  x_train, x_test, y_train, y_test = x[train_index], x[test_index], y[train_index], y[test_index]
  clf.fit(x_train, y_train)
  pred = clf.predict(x_test)
  accuracies.append(metrics.accuracy_score(y_test, pred))
  precisions.append(metrics.precision_score(y_test, pred))
  recalls.append(metrics.recall_score(y_test, pred))
print("Accuracy:",mean(accuracies))
print("Precision:",mean(precisions))
print("Recall:",mean(recalls))
overall_acc.append(mean(accuracies) * 100)
overall_prec.append(mean(precisions) * 100)
overall_rec.append(mean(recalls) * 100)
sc.append("Unbalanced-3-F-overlap")

Accuracy: 1.0
Precision: 1.0
Recall: 1.0


5.6: **Balanced dataset with all feature overlap**

In [ ]:
accuracies = []
precisions = []
recalls = []

features = ['X1','X2','X3','X4']
x = Balanced_AllF_overlap.loc[:, features].values
x = pd.DataFrame(data = x, columns = features)
y = Balanced_AllF_overlap.Y
scaler = MinMaxScaler(feature_range=(0, 1))
x = scaler.fit_transform(x)

In [ ]:
clf = svm.SVC()
for train_index, test_index in Kfold.split(x):
  x_train, x_test, y_train, y_test = x[train_index], x[test_index], y[train_index], y[test_index]
  clf.fit(x_train, y_train)
  pred = clf.predict(x_test)
  accuracies.append(metrics.accuracy_score(y_test, pred))
  precisions.append(metrics.precision_score(y_test, pred))
  recalls.append(metrics.recall_score(y_test, pred))
print("Accuracy:",mean(accuracies))
print("Precision:",mean(precisions))
print("Recall:",mean(recalls))
overall_acc.append(mean(accuracies) * 100)
overall_prec.append(mean(precisions) * 100)
overall_rec.append(mean(recalls) * 100)
sc.append("Balanced-all-F-overlap")

Accuracy: 0.8066666666666669
Precision: 0.7388466431229588
Recall: 0.9467565359477124


5.7: **Unbalanced dataset with all feature overlap**

In [ ]:
accuracies = []
precisions = []
recalls = []

features = ['X1','X2','X3','X4']
x = unbalanced_allF_overlap.loc[:, features].values
x = pd.DataFrame(data = x, columns = features)
y = unbalanced_allF_overlap.Y
scaler = MinMaxScaler(feature_range=(0, 1))
x = scaler.fit_transform(x)

In [ ]:
clf = svm.SVC()
for train_index, test_index in Kfold.split(x):
  x_train, x_test, y_train, y_test = x[train_index], x[test_index], y[train_index], y[test_index]
  clf.fit(x_train, y_train)
  pred = clf.predict(x_test)
  accuracies.append(metrics.accuracy_score(y_test, pred))
  precisions.append(metrics.precision_score(y_test, pred))
  recalls.append(metrics.recall_score(y_test, pred))
print("Accuracy:",mean(accuracies))
print("Precision:",mean(precisions))
print("Recall:",mean(recalls))
overall_acc.append(mean(accuracies) * 100)
overall_prec.append(mean(precisions) * 100)
overall_rec.append(mean(recalls) * 100)
sc.append("Unbalanced-all-F-overlap")

Accuracy: 0.9533333333333334
Precision: 0.9589180602006688
Recall: 0.9827898550724636


5.8: **Balanced dataset with no feature overlap**

In [ ]:
accuracies = []
precisions = []
recalls = []

features = [' X1','X2','X3','X4']
x = Balanced_noF_overlap.loc[:, features].values
x = pd.DataFrame(data = x, columns = features)
y = Balanced_noF_overlap.Y
scaler = MinMaxScaler(feature_range=(0, 1))
x = scaler.fit_transform(x)

In [ ]:
clf = svm.SVC()
for train_index, test_index in Kfold.split(x):
  x_train, x_test, y_train, y_test = x[train_index], x[test_index], y[train_index], y[test_index]
  clf.fit(x_train, y_train)
  pred = clf.predict(x_test)
  accuracies.append(metrics.accuracy_score(y_test, pred))
  precisions.append(metrics.precision_score(y_test, pred))
  recalls.append(metrics.recall_score(y_test, pred))
print("Accuracy:",mean(accuracies))
print("Precision:",mean(precisions))
print("Recall:",mean(recalls))
overall_acc.append(mean(accuracies) * 100)
overall_prec.append(mean(precisions) * 100)
overall_rec.append(mean(recalls) * 100)
sc.append("Balanced-no-F-overlap")

Accuracy: 1.0
Precision: 1.0
Recall: 1.0


5.9: **Unbalanced dataset with no feature overlap**

In [ ]:
accuracies = []
precisions = []
recalls = []

features = [' X1','X2','X3','X4']
x = unbalanced_noF_overlap.loc[:, features].values
x = pd.DataFrame(data = x, columns = features)
y = unbalanced_noF_overlap.Y
scaler = MinMaxScaler(feature_range=(0, 1))
x = scaler.fit_transform(x)

In [ ]:
clf = svm.SVC()
for train_index, test_index in Kfold.split(x):
  x_train, x_test, y_train, y_test = x[train_index], x[test_index], y[train_index], y[test_index]
  clf.fit(x_train, y_train)
  pred = clf.predict(x_test)
  accuracies.append(metrics.accuracy_score(y_test, pred))
  precisions.append(metrics.precision_score(y_test, pred))
  recalls.append(metrics.recall_score(y_test, pred))
print("Accuracy:",mean(accuracies))
print("Precision:",mean(precisions))
print("Recall:",mean(recalls))
overall_acc.append(mean(accuracies) * 100)
overall_prec.append(mean(precisions) * 100)
overall_rec.append(mean(recalls) * 100)
sc.append("Unbalanced-no-F-overlap")

Accuracy: 1.0
Precision: 1.0
Recall: 1.0


Tabulating the results,

In [ ]:
SVM_Res = pd.DataFrame(list(zip(sc, overall_acc, overall_prec, overall_rec)), columns = ['SVM - Dataset nature','Accuracy%','Precision%','Recall%'])
SVM_Res

SVM - Dataset nature   Accuracy%  Precision%     Recall%
0      Balanced-1-F-overlap  100.000000  100.000000  100.000000
1    Unbalanced-1-F-overlap  100.000000  100.000000  100.000000
2      Balanced-2-F-overlap  100.000000  100.000000  100.000000
3      Balanced-3-F-overlap  100.000000  100.000000  100.000000
4    Unbalanced-3-F-overlap  100.000000  100.000000  100.000000
5    Balanced-all-F-overlap   80.666667   73.884664   94.675654
6  Unbalanced-all-F-overlap   95.333333   95.891806   98.278986
7     Balanced-no-F-overlap  100.000000  100.000000  100.000000
8   Unbalanced-no-F-overlap  100.000000  100.000000  100.000000

6. **Running Decision tree classifier using 10-fold cross validation for balanced and unbalanced datasets with varying degrees of overlapping**

In [ ]:
Kfold = KFold(10, True)


6.1: **Balanced dataset with 1 feature overlap**

6.2: **Unbalanced dataset with 1 feature overlap**

6.3: **Balanced dataset with 2 feature overlap**

6.4: **Balanced dataset with 3 feature overlap**

6.5: **Unbalanced dataset with 3 feature overlap**

6.6: **Balanced dataset with all feature overlap**

6.7: **Unbalanced dataset with all feature overlap**

6.8: **Balanced dataset with no feature overlap**

6.9: **Unbalanced dataset with no feature overlap**

Tabulating the results,

7. **Running KNN (k = 3) classifier using 10-fold cross validation for balanced and unbalanced datasets with varying degrees of overlapping**

In [ ]:
Kfold = KFold(10, True)

7.1: **Balanced dataset with 1 feature overlap**

7.2: **Unbalanced dataset with 1 feature overlap**

7.3: **Balanced dataset with 2 feature overlap**

7.4: **Balanced dataset with 3 feature overlap**

7.5: **Unbalanced dataset with 3 feature overlap**

7.6: **Balanced dataset with all feature overlap**

7.7: **Unbalanced dataset with all feature overlap**

7.8: **Balanced dataset with no feature overlap**

7.9: **Unbalanced dataset with no feature overlap**

Tabulating the results,

8. **Oversampling an unbalanced dataset with all features overlapping using ADASYN
algorithm, before running all three classifiers**

9. **Undersampling the majority class in an unbalanced dataset with all features overlapping
using Random Under Sampler algorithm, before running all three classifiers.**

**Inference:**